juniper_graphql_ws/src/lib.rs

```diff
@@ -0,0 +1,1071 @@
+/// message.
+pub struct ConnectionConfig<CtxT> {
+    context: CtxT,
+    max_in_flight_operations: usize,
+    keep_alive_interval: Duration,
+}
+
+impl<CtxT> ConnectionConfig<CtxT> {
+    /// Constructs the configuration required for a connection to be accepted.
+    pub fn new(context: CtxT) -> Self {
+        Self {
+            context,
+            max_in_flight_operations: 0,
+            keep_alive_interval: Duration::from_secs(30),
+        }
+    }
+
+    /// Specifies the maximum number of in-flight operations that a connection can have. If this
+    /// number is exceeded, attempting to start more will result in an error. By default, there is
+    /// no limit to in-flight operations.
+    pub fn with_max_in_flight_operations(mut self, max: usize) -> Self {
+        self.max_in_flight_operations = max;
+        self
+    }
+
+    /// Specifies the interval at which to send keep-alives. Specifying a zero duration will
+    /// disable keep-alives. By default, keep-alives are sent every
+    /// 30 seconds.
+    pub fn with_keep_alive_interval(mut self, interval: Duration) -> Self {
+        self.keep_alive_interval = interval;
+        self
+    }
+}
+
+impl<S: ScalarValue, CtxT: Unpin + Send + 'static> Init<S, CtxT> for ConnectionConfig<CtxT> {
+    type Error = Infallible;
+    type Future = future::Ready<Result<Self, Self::Error>>;
+
+    fn init(self, _params: Variables<S>) -> Self::Future {
+        future::ready(Ok(self))
+    }
+}
+
+enum Reaction<S: Schema> {
+    ServerMessage(ServerMessage<S::ScalarValue>),
+    EndStream,
+}
+
+impl<S: Schema> Reaction<S> {
+    /// Converts the reaction into a one-item stream.
+    fn to_stream(self) -> BoxStream<'static, Self> {
+        stream::once(future::ready(self)).boxed()
+    }
+}
+
+/// Init defines the requirements for types that can provide connection configurations when
+/// ConnectionInit messages are received. Implementations are provided for `ConnectionConfig` and
+/// closures that meet the requirements.
+pub trait Init<S: ScalarValue, CtxT>: Unpin + 'static {
+    /// The error that is returned on failure. The formatted error will be used as the contents of
+    /// the "message" field sent back to the client.
+    type Error: Error;
+
+    /// The future configuration type.
+    type Future: Future<Output = Result<ConnectionConfig<CtxT>, Self::Error>> + Send + 'static;
+
+    /// Returns a future for the configuration to use.
+    fn init(self, params: Variables<S>) -> Self::Future;
+}
+
+impl<F, S, CtxT, Fut, E> Init<S, CtxT> for F
+where
+    S: ScalarValue,
+    F: FnOnce(Variables<S>) -> Fut + Unpin + 'static,
+    Fut: Future<Output = Result<ConnectionConfig<CtxT>, E>> + Send + 'static,
+    E: Error,
+{
+    type Error = E;
+    type Future = Fut;
+
+    fn init(self, params: Variables<S>) -> Fut {
+        self(params)
+    }
+}
+
+enum ConnectionState<S: Schema, I: Init<S::ScalarValue, S::Context>> {
```

Comment:
This seems like a great opportunity to use the typestate pattern to make sure all state transitions are valid and checked at compile-time:

http://cliffle.com/blog/rust-typestate/#:~:text=The%20typestate%20pattern%20is%20an,in%20its%20compile-time%20type.

Reply:
Hmm... I experimented with a few things and couldn't figure out how to increase compiler coverage while keeping this elegant. The pre-init state effectively contains a `FnOnce`, which already gives us just about all the compile-time guarantees we can make. In order to invoke that, we _have_ to consume the state and move to a different one, of which all are valid.